src/robot/devices/core/lsm6/lsm6.ts

```diff
@@ -202,141 +301,307 @@ export default class LSM6 {
+     * Perform a read of the FIFO buffer and store any saved frames into our
+     * internal storage.
      *
-     * The values obtained from the raw I2C read operations are the
-     * 16-bit twos-complement raw IMU values. They will need to be
-     * multiplied by the sensitivity values to obtain a reading in mDPS.
-     *
-     * The final values that get stored are in degrees-per-second (dps)
-     *
-     * Angular Rate Sensitivity: G_So (page 15)
+     * A "frame" consists of a snapshot of gyro and accelerometer values. The
+     * time between each "frame" is dependent on the FIFO ODR value
+     * @private
      */
-    public async readGyro(): Promise<void> {
-        if (!this._isReady) {
-            return;
+    private async _fifoLoop(): Promise<void> {
+        const start = Date.now();
+        const status = await this._fifoGetStatus();
+        logger.silly(`FIFO Status: 0x${status.toString(16)}`);
+
+        const numUnread = status & 0xFFF;
+        logger.silly(`Num unread entries: ${numUnread}`);
+        const valBuf = Buffer.alloc(2);
+
+        const tempFrame: FIFOFrame = {
+            gyroX: 0,
+            gyroY: 0,
+            gyroZ: 0,
+            accelX: 0,
+            accelY: 0,
+            accelZ: 0
         }
 
-        const xlg = await this._readByte(RegAddr.OUTX_L_G);
-        const xhg = await this._readByte(RegAddr.OUTX_H_G);
-        const ylg = await this._readByte(RegAddr.OUTY_L_G);
-        const yhg = await this._readByte(RegAddr.OUTY_H_G);
-        const zlg = await this._readByte(RegAddr.OUTZ_L_G);
-        const zhg = await this._readByte(RegAddr.OUTZ_H_G);
-
-        const gyroX = (xhg << 8) | xlg;
-        const gyroY = (yhg << 8) | ylg;
-        const gyroZ = (zhg << 8) | zlg;
-
-        const tmpBuf = Buffer.alloc(2);
+        const accelScaleFactor = ACCEL_OUTPUT_SCALE_FACTOR.get(this._settings.accelRange);
+        const gyroScaleFactor = GYRO_OUTPUT_SCALE_FACTOR.get(this._settings.gyroRange);
+
+        // We only want to process groups of 6 values (which correspond to 3 gyro + 3 accel values)
+        // An IMU "data frame" consists of 6 reads off the FIFO register
+        // The order of values read are: GyroX, GyroY, GyroZ, AccelX, AccelY, AccelZ
+        //
+        // All values read from the FIFO are read in as 16-bit twos-complement raw values.
+        // These need to be multiplied by the sensitivity values to obtain a reading in
+        // mG (for the accelerometer) or mDPS (for the gyro). Frame data is stored with
+        // the following units:
+        // G for accelerometer values
+        // DPS for gyro values
+        //
+        // References
+        // Angular Rate Sensitivity: G_So (page 15)
+        // Linear Acceleration Sensitivity: LA_So (page 15)
+        const shouldProcess = (numUnread % 6 === 0);
+        for (let i = 0; i < numUnread; i++) {
+            let val = await this._readWord(RegAddr.FIFO_DATA_OUT_L);
```

Comment:
Are you leveraging the fact that registers are in sequence, FIFO_DATA_OUT_L is 0x3E and FIFO_DATA_OUT_H is 0x3F, to read as a word?  I was expecting to see readByte's for each. Might be good to note here if that's the case. Or maybe I'm misinterpreting the datasheet.

Reply:
Yeah, I can add a note